cmd/pi-strings-go/strings.go

```diff
@@ -77,3 +79,72 @@ func findUTF8String(buf []byte, minLength int) (s string, n uint64, ok bool) {
 	}
 	return string(buf[:n]), n, true
 }
+
+// extractUTF16LittleEndianStrings extracts printable UTF-16 strings in little
+// endian byte order with a minimum number of characters from the given binary.
+func extractUTF16LittleEndianStrings(buf []byte, minLength int, c chan []*stringspb.StringInfo) {
+	extractUTF16Strings(buf, minLength, binary.LittleEndian, c)
+}
+
+// extractUTF16BigEndianStrings extracts printable UTF-16 strings in big endian
+// byte order with a minimum number of characters from the given binary.
+func extractUTF16BigEndianStrings(buf []byte, minLength int, c chan []*stringspb.StringInfo) {
+	extractUTF16Strings(buf, minLength, binary.BigEndian, c)
+}
+
+// extractUTF16Strings extracts printable UTF-8 strings with a minimum number of
+// characters from the given binary. The given byte order is used to decode
+// uint16 values of buf.
+func extractUTF16Strings(buf []byte, minLength int, order binary.ByteOrder, c chan []*stringspb.StringInfo) {
+	var infos []*stringspb.StringInfo
+	for i := 0; i < len(buf); {
+		start := uint64(i)
+		s, n, ok := findUTF16String(buf[start:], minLength, order)
+		i += int(n)
+		if !ok {
+			continue
+		}
+		info := &stringspb.StringInfo{
+			Location:  start,
+			RawString: s,
+			Size:      n,
+			Encoding:  stringspb.Encoding_UTF16,
+		}
+		infos = append(infos, info)
+	}
+	c <- infos
+}
+
+// findUTF16String tries to locate the longest printable UTF-8 string starting
+// at buf. For the string to be valid, it must be of at least the specified
+// minimum length in number of characters. The given byte order is used to
+// decode uint16 values of buf. The integer return value n specifies the number
+// of bytes read, and the boolean return value indicates success. If an invalid
+// UTF-8 encoding is encountered at the start of the given buffer, n is set to 1
+// and the boolean return value is false.
+func findUTF16String(buf []byte, minLength int, order binary.ByteOrder) (s string, n uint64, ok bool) {
+	n = 0
+	nchars := 0
+	s = ""
+	for i := 0; i < len(buf); {
+		r, size := utf16DecodeRuneWithOrder(buf[i:], order)
+		if r == unicode.ReplacementChar {
+			// invalid UTF-16.
+			break
+		}
+		if !unicode.IsGraphic(r) {
+			// non-printable char.
+			break
+		}
+		// found valid char.
+		i += size
+		n = uint64(i)
+		nchars++
+		s += string(r) // TODO: check if we are allowed to convert runes to strings like this.
```

Comment:
Hard to check what the string convert does under the hood. 

Found this Chinese site which solves it with a one-liner: https://angelonotes.blogspot.com/2015/09/golang-utf16-utf8.html